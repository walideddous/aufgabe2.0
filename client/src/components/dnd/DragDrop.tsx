import React, { useState, useCallback } from "react";
import { DragDropContext, Droppable, Draggable } from "react-beautiful-dnd";
import { Card, Col } from "antd";
import { DeleteOutlined } from "@ant-design/icons";

// Import types
import { TstateDND, Tchoose, Tstations } from "../type/Types";

// Declare Types
interface TporpsDND {
  choose: Tchoose;
  stateDND: TstateDND;
  selected: Tstations | undefined;
  lastAutoSelectElem: Tstations | undefined;
  onclick: (e: { id: string | number; name: string }, index: number) => void;
  handleAddStopsOnCLick: (
    e: { id: string | number; name: string },
    index: number
  ) => void;
  onDelete: (
    e: { id: string | number; name: string },
    SourceOrTarget: string,
    index: number
  ) => void;
  handleDragEnd: (e: any) => void;
}

const DragDrop = ({
  choose,
  stateDND,
  selected,
  lastAutoSelectElem,
  handleAddStopsOnCLick,
  handleDragEnd,
  onclick,
  onDelete,
}: TporpsDND) => {
  const [clicked, setClicked] = useState(false);

  const handleClick = useCallback(
    (e: { id: string | number; name: string }, index: number) => {
      onclick(e, index);
      setClicked(!clicked);
    },
    [onclick, setClicked, clicked]
  );

  const handleDelete = useCallback(
    (
      e: { id: string | number; name: string },
      SourceOrTarget: string,
      index: number
    ) => {
      onDelete(e, SourceOrTarget, index);
    },
    [onDelete]
  );

  const addStopsOnCLick = useCallback(
    (el: { id: string | number; name: string }, index: number) => {
      handleAddStopsOnCLick(el, index);
    },
    [handleAddStopsOnCLick]
  );

  return (
    <div className="App">
      <DragDropContext onDragEnd={handleDragEnd}>
        <Col span={12}>
          <Card bordered={true} title={stateDND.vorschlag.title}>
            <Droppable droppableId={"vorschlag"}>
              {(provided: any) => {
                return (
                  <div
                    ref={provided.innerRef}
                    {...provided.droppableProps}
                    className={"droppable-col"}
                  >
                    {stateDND.vorschlag.items.map((el: any, index: number) => {
                      return (
                        <Draggable
                          key={el.id}
                          index={index}
                          draggableId={el.id}
                        >
                          {(provided) => {
                            return (
                              <div
                                className="item-suggestion "
                                ref={provided.innerRef}
                                {...provided.draggableProps}
                                {...provided.dragHandleProps}
                              >
                                <span
                                  style={{ width: "90%", height: "30px" }}
                                  onClick={() => {
                                    addStopsOnCLick(el, index);
                                  }}
                                >
                                  {el.name}
                                </span>
                                {/*<Button
                                  type="dashed"
                                  shape="round"
                                  icon={<DeleteOutlined />}
                                  onClick={() => {
                                    handleDelete(
                                      el,
                                      stateDND.vorschlag.title,
                                      index
                                    );
                                  }}
                                />*/}
                              </div>
                            );
                          }}
                        </Draggable>
                      );
                    })}
                    {provided.placeholder}
                  </div>
                );
              }}
            </Droppable>
          </Card>
        </Col>
        <Col span={12}>
          <Card bordered={true} title={stateDND.trajekt.title}>
            <Droppable droppableId={"trajekt"}>
              {(provided: any) => {
                return (
                  <div
                    ref={provided.innerRef}
                    {...provided.droppableProps}
                    className={"droppable-col"}
                  >
                    {stateDND.trajekt.items.map((el: any, index: number) => {
                      return (
                        <Draggable
                          key={el.id}
                          index={index}
                          draggableId={el.id}
                        >
                          {(provided) => {
                            return (
                              <div
                                className={
                                  (lastAutoSelectElem &&
                                    !selected &&
                                    stateDND.trajekt.items.length - 1 ===
                                      index) ||
                                  (lastAutoSelectElem &&
                                    selected &&
                                    selected.name === el.name) ||
                                  (selected &&
                                    !lastAutoSelectElem &&
                                    selected.name === el.name)
                                    ? "item-highlighted"
                                    : "item"
                                }
                                ref={provided.innerRef}
                                {...provided.draggableProps}
                                {...provided.dragHandleProps}
                              >
                                <span
                                  onClick={() => {
                                    handleClick(el, index);
                                  }}
                                  style={{ width: "90%" }}
                                >
                                  {el.name}
                                </span>
                                <button
                                  style={{
                                    backgroundColor: "white",
                                    color: "#3949ab",
                                    borderRadius: "5px",
                                    outline: "0",
                                    cursor: "pointer",
                                    boxShadow: "0px 2px 2px lightgray",
                                  }}
                                  onClick={() => {
                                    handleDelete(
                                      el,
                                      stateDND.trajekt.title,
                                      index
                                    );
                                  }}
                                >
                                  <DeleteOutlined />
                                </button>
                              </div>
                            );
                          }}
                        </Draggable>
                      );
                    })}
                    {provided.placeholder}
                  </div>
                );
              }}
            </Droppable>
          </Card>
        </Col>
      </DragDropContext>
    </div>
  );
};

export default React.memo(DragDrop);
