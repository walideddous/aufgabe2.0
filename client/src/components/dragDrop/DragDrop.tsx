import React, { useState, useEffect, Fragment } from "react";
import { DragDropContext, Droppable, Draggable } from "react-beautiful-dnd";
import { Card, Col, Button } from "antd";
import { DeleteOutlined, ArrowUpOutlined } from "@ant-design/icons";

// Typescript
import { Tstations, TstateDND } from "../../types/types";

// Declare Types
interface TporpsDND {
  stateDND: TstateDND;
  selected: Tstations | undefined;
  onResetStopSequence: () => void;
  onAddStopsOnCLick: (e: any) => void;
  onDragEnd: (e: any) => void;
  onClickOnDrop: (e: any, index: number) => void;
  onDeleteDND: (e: any, index: number) => void;
}

const DragDrop = ({
  stateDND,
  selected,
  onResetStopSequence,
  onAddStopsOnCLick,
  onDragEnd,
  onClickOnDrop,
  onDeleteDND,
}: TporpsDND) => {
  const [clicked, setClicked] = useState<boolean>(false);
  const [hide, setHide] = useState<boolean>(false);

  const resize = () => {
    //@ts-ignore
    if (window.innerWidth < 992) {
      setHide(true);
    } else {
      setHide(false);
    }
  };

  useEffect(() => {
    window.addEventListener("resize", resize);
    resize();
    return () => {
      window.removeEventListener("resize", resize);
    };
  });

  return (
    <Fragment>
      <DragDropContext onDragEnd={onDragEnd}>
        <Col xs={24}>
          <Card title={stateDND.suggestions.title}>
            <Droppable droppableId={"suggestions"}>
              {(provided: any) => {
                return (
                  <div
                    style={{
                      display: "flex",
                    }}
                  >
                    <Col lg={12} xs={24}>
                      <div
                        ref={provided.innerRef}
                        {...provided.droppableProps}
                        className={"droppable-col"}
                      >
                        {stateDND.suggestions.items &&
                          stateDND.suggestions.items
                            .slice(0, 8)
                            .map((el: any, index: number) => {
                              return (
                                <Draggable
                                  key={el._id}
                                  index={index}
                                  draggableId={el._id + index}
                                >
                                  {(provided) => {
                                    return (
                                      <div
                                        className="item-suggestion"
                                        ref={provided.innerRef}
                                        {...provided.draggableProps}
                                        {...provided.dragHandleProps}
                                      >
                                        <span
                                          id="addStopsButton"
                                          style={{ width: "90%" }}
                                          onClick={() => {
                                            onAddStopsOnCLick(el);
                                          }}
                                        >
                                          {el.name} "{el.distance.toFixed(3)}{" "}
                                          Km"
                                        </span>
                                        <button
                                          style={{
                                            width: "30px",
                                            backgroundColor: "white",
                                            color: "#3949ab",
                                            borderRadius: "5px",
                                            outline: "0",
                                          }}
                                        >
                                          <ArrowUpOutlined
                                            style={{
                                              transform: `rotate(${el.angle}deg)`,
                                            }}
                                          />
                                        </button>
                                      </div>
                                    );
                                  }}
                                </Draggable>
                              );
                            })}
                        {provided.placeholder}
                      </div>
                    </Col>
                    {!hide ? (
                      <Col lg={12}>
                        <div
                          ref={provided.innerRef}
                          {...provided.droppableProps}
                          className={"droppable-col"}
                        >
                          {stateDND.suggestions.items &&
                            stateDND.suggestions.items
                              .slice(8, 16)
                              .map((el: any, index: number) => {
                                return (
                                  <Draggable
                                    key={el._id}
                                    index={index + 8}
                                    draggableId={el._id + index}
                                  >
                                    {(provided) => {
                                      return (
                                        <div
                                          className="item-suggestion "
                                          ref={provided.innerRef}
                                          {...provided.draggableProps}
                                          {...provided.dragHandleProps}
                                        >
                                          <span
                                            id="addStopsButton"
                                            style={{
                                              width: "90%",
                                            }}
                                            onClick={() => {
                                              onAddStopsOnCLick(el);
                                            }}
                                          >
                                            {el.name} "{el.distance.toFixed(3)}{" "}
                                            Km"
                                          </span>
                                          <button
                                            style={{
                                              width: "30px",
                                              backgroundColor: "white",
                                              color: "#3949ab",
                                              borderRadius: "5px",
                                              outline: "0",
                                            }}
                                          >
                                            <ArrowUpOutlined
                                              style={{
                                                transform: `rotate(${el.angle}deg)`,
                                              }}
                                            />
                                          </button>
                                        </div>
                                      );
                                    }}
                                  </Draggable>
                                );
                              })}
                          {provided.placeholder}
                        </div>
                      </Col>
                    ) : (
                      <div
                        ref={provided.innerRef}
                        {...provided.droppableProps}
                      ></div>
                    )}
                  </div>
                );
              }}
            </Droppable>
          </Card>
        </Col>
        <Col xs={24}>
          <Card
            title={stateDND.trajekt.title}
            extra={
              stateDND.trajekt.items.length ? (
                <Button
                  id="clearAll_button"
                  onClick={() => onResetStopSequence()}
                >
                  Reset
                </Button>
              ) : null
            }
          >
            <div
              style={
                stateDND.trajekt.items.length > 0 &&
                stateDND.trajekt.items.length < 7
                  ? {
                      height: "370px",
                      overflowY: "auto",
                    }
                  : stateDND.trajekt.items.length >= 7
                  ? {
                      height: "370px",
                      overflowY: "auto",
                      display: "flex",
                      flexDirection: "column-reverse",
                    }
                  : {}
              }
            >
              <Droppable droppableId={"trajekt"}>
                {(provided: any) => {
                  return (
                    <div
                      ref={provided.innerRef}
                      {...provided.droppableProps}
                      className={"droppable-col"}
                    >
                      {stateDND.trajekt.items &&
                        stateDND.trajekt.items.map((el: any, index: number) => {
                          return (
                            <Draggable
                              key={el._id + index}
                              index={index}
                              draggableId={el._id + index}
                            >
                              {(provided) => {
                                return (
                                  <div
                                    id={
                                      selected &&
                                      selected._id + selected.index ===
                                        el._id + index
                                        ? "item-highlighted"
                                        : "item"
                                    }
                                    className={
                                      selected &&
                                      selected._id + selected.index ===
                                        el._id + index
                                        ? "item-highlighted"
                                        : "item"
                                    }
                                    ref={provided.innerRef}
                                    {...provided.draggableProps}
                                    {...provided.dragHandleProps}
                                  >
                                    <span
                                      id="clickStops"
                                      onClick={() => {
                                        onClickOnDrop(el, index);
                                        setClicked(!clicked);
                                      }}
                                      style={{ width: "90%" }}
                                    >
                                      {el.name}
                                    </span>
                                    <button
                                      id="deleteStopButton"
                                      style={{
                                        width: "30px",
                                        backgroundColor: "white",
                                        color: "#3949ab",
                                        borderRadius: "5px",
                                        outline: "0",
                                        cursor: "pointer",
                                        boxShadow: "0px 2px 2px lightgray",
                                      }}
                                      onClick={() => {
                                        onDeleteDND(el, index);
                                      }}
                                    >
                                      <DeleteOutlined />
                                    </button>
                                  </div>
                                );
                              }}
                            </Draggable>
                          );
                        })}
                      {provided.placeholder}
                    </div>
                  );
                }}
              </Droppable>
            </div>
          </Card>
        </Col>
      </DragDropContext>
    </Fragment>
  );
};

export default React.memo(DragDrop);
