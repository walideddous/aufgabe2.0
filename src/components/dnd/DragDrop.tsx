import React, { useState } from "react";
import { DragDropContext, Droppable, Draggable } from "react-beautiful-dnd";
import { Card, Col, Button } from "antd";
import { DeleteOutlined } from "@ant-design/icons";

// Import types
import { TstateDND, Tchoose } from "../type/Types";

// Declare Types
interface TporpsDND {
  choose: Tchoose;
  stateDND: TstateDND;
  onclick: (e: { id: string | number; name: string }) => void;
  onDelete: (e: { id: string | number; name: string }) => void;
  handleDragEnd: (e: any) => void;
}

const DragDrop = ({
  choose,
  stateDND,
  handleDragEnd,
  onclick,
  onDelete,
}: TporpsDND) => {
  const [clicked, setClicked] = useState(false);

  const handleClick = (e: { id: string | number; name: string }) => {
    onclick(e);
    setClicked(!clicked);
  };

  const handleDelete = (e: { id: string | number; name: string }) => {
    onDelete(e);
  };

  // Handle the highlitimg
  const handleButtonColor = () => {};

  return (
    <div className="App">
      <DragDropContext onDragEnd={handleDragEnd}>
        <Col span={12}>
          <Card bordered={true} title={stateDND.vorschlag.title}>
            <Droppable droppableId={"vorschlag"}>
              {(provided: any) => {
                return (
                  <div
                    ref={provided.innerRef}
                    {...provided.droppableProps}
                    className={"droppable-col"}
                  >
                    {stateDND.vorschlag.items.map((el: any, index: number) => {
                      return (
                        <Draggable
                          key={el.id}
                          index={index}
                          draggableId={el.id}
                        >
                          {(provided) => {
                            return (
                              <div
                                className="item"
                                ref={provided.innerRef}
                                {...provided.draggableProps}
                                {...provided.dragHandleProps}
                              >
                                <span
                                  onClick={() => {
                                    handleClick(el);
                                  }}
                                >
                                  {el.name}
                                </span>
                                <Button
                                  type="dashed"
                                  shape="circle"
                                  icon={<DeleteOutlined />}
                                  onClick={() => {
                                    handleDelete(el);
                                  }}
                                />
                              </div>
                            );
                          }}
                        </Draggable>
                      );
                    })}
                    {provided.placeholder}
                  </div>
                );
              }}
            </Droppable>
          </Card>
        </Col>
        <Col span={12}>
          <Card bordered={true} title={stateDND.trajekt.title}>
            <Droppable droppableId={"trajekt"}>
              {(provided: any) => {
                return (
                  <div
                    ref={provided.innerRef}
                    {...provided.droppableProps}
                    className={"droppable-col"}
                  >
                    {stateDND.trajekt.items.map((el: any, index: number) => {
                      return (
                        <Draggable
                          key={el.id}
                          index={index}
                          draggableId={el.id}
                        >
                          {(provided) => {
                            return (
                              <div
                                className={
                                  stateDND.trajekt.items[
                                    stateDND.trajekt.items.length - 1
                                  ].id === el.id
                                    ? "item-highlighted"
                                    : "item"
                                }
                                ref={provided.innerRef}
                                {...provided.draggableProps}
                                {...provided.dragHandleProps}
                              >
                                <span
                                  onClick={() => {
                                    handleClick(el);
                                  }}
                                >
                                  {el.name}
                                </span>
                                <Button
                                  type="dashed"
                                  shape="circle"
                                  icon={<DeleteOutlined />}
                                  onClick={() => {
                                    handleDelete(el);
                                  }}
                                />
                              </div>
                            );
                          }}
                        </Draggable>
                      );
                    })}
                    {provided.placeholder}
                  </div>
                );
              }}
            </Droppable>
          </Card>
        </Col>
      </DragDropContext>
    </div>
  );
};

export default DragDrop;
