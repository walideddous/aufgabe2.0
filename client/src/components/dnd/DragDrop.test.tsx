import React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import DragDrop from "./DragDrop";

const setUp = (props: any) => {
  const component = mount(<DragDrop {...props} />);
  return component;
};

describe("DragDrop component", () => {
  let mountWrapper: any;
  let props = {
    stateDND: {
      vorschlag: {
        title: "Suggestion",
        items: [],
      },
      trajekt: {
        title: "Stop sequence",
        items: [],
      },
    },
    selected: [],
    handleDragEnd: jest.fn(),
    handleAddStopsOnCLick: jest.fn(),
    onclick: jest.fn(),
    onDelete: jest.fn(),
  };

  beforeEach(() => {
    mountWrapper = setUp(props);
  });

  it("match snapshot with the DrapDrop component", () => {
    expect(toJSON(mountWrapper)).toMatchSnapshot();
  });
});
