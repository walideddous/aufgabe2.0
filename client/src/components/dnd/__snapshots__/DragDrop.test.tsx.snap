// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragDrop component match snapshot with the DrapDrop component 1`] = `
<Memo(DragDrop)
  handleAddStopsOnCLick={[MockFunction]}
  handleDragEnd={[MockFunction]}
  onDelete={[MockFunction]}
  onclick={[MockFunction]}
  selected={Array []}
  stateDND={
    Object {
      "trajekt": Object {
        "items": Array [],
        "title": "Stop sequence",
      },
      "vorschlag": Object {
        "items": Array [],
        "title": "Suggestion",
      },
    }
  }
>
  <DragDropContext
    onDragEnd={[MockFunction]}
  >
    <ErrorBoundary>
      <App
        contextId="0"
        dragHandleUsageInstructions="
  Press space bar to start a drag.
  When dragging you can use the arrow keys to move the item around and escape to cancel.
  Some screen readers may require you to be in focus mode or to use your pass through key
"
        onDragEnd={[MockFunction]}
        setCallbacks={[Function]}
      >
        <Provider
          context={
            Object {
              "$$typeof": Symbol(react.context),
              "Consumer": Object {
                "$$typeof": Symbol(react.context),
                "_calculateChangedBits": null,
                "_context": [Circular],
              },
              "Provider": Object {
                "$$typeof": Symbol(react.provider),
                "_context": [Circular],
              },
              "_calculateChangedBits": null,
              "_currentRenderer": Object {},
              "_currentRenderer2": null,
              "_currentValue": null,
              "_currentValue2": null,
              "_threadCount": 0,
            }
          }
          store={
            Object {
              "dispatch": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
              Symbol(observable): [Function],
            }
          }
        >
          <Col
            xs={24}
          >
            <div
              className="ant-col ant-col-xs-24"
              style={Object {}}
            >
              <Card
                bordered={true}
                title="Suggestion"
              >
                <div
                  className="ant-card ant-card-bordered"
                >
                  <div
                    className="ant-card-head"
                    style={Object {}}
                  >
                    <div
                      className="ant-card-head-wrapper"
                    >
                      <div
                        className="ant-card-head-title"
                      >
                        Suggestion
                      </div>
                    </div>
                  </div>
                  <div
                    className="ant-card-body"
                    style={Object {}}
                  >
                    <ConnectFunction
                      direction="vertical"
                      droppableId="vorschlag"
                      getContainerForClone={[Function]}
                      ignoreContainerClipping={false}
                      isCombineEnabled={false}
                      isDropDisabled={false}
                      mode="standard"
                      renderClone={null}
                      type="DEFAULT"
                    >
                      <Droppable
                        direction="vertical"
                        droppableId="vorschlag"
                        getContainerForClone={[Function]}
                        ignoreContainerClipping={false}
                        isCombineEnabled={false}
                        isDropDisabled={false}
                        mode="standard"
                        placeholder={null}
                        renderClone={null}
                        shouldAnimatePlaceholder={false}
                        snapshot={
                          Object {
                            "draggingFromThisWith": null,
                            "draggingOverWith": null,
                            "isDraggingOver": false,
                            "isUsingPlaceholder": false,
                          }
                        }
                        type="DEFAULT"
                        updateViewportMaxScroll={[Function]}
                        useClone={null}
                      >
                        <div
                          style={
                            Object {
                              "display": "flex",
                            }
                          }
                        >
                          <Col
                            lg={12}
                            xs={24}
                          >
                            <div
                              className="ant-col ant-col-xs-24 ant-col-lg-12"
                              style={Object {}}
                            >
                              <div
                                className="droppable-col"
                                data-rbd-droppable-context-id="0"
                                data-rbd-droppable-id="vorschlag"
                              >
                                <AnimateInOut
                                  on={null}
                                  shouldAnimate={false}
                                />
                              </div>
                            </div>
                          </Col>
                          <Col
                            lg={12}
                          >
                            <div
                              className="ant-col ant-col-lg-12"
                              style={Object {}}
                            >
                              <div
                                className="droppable-col"
                                data-rbd-droppable-context-id="0"
                                data-rbd-droppable-id="vorschlag"
                              >
                                <AnimateInOut
                                  on={null}
                                  shouldAnimate={false}
                                />
                              </div>
                            </div>
                          </Col>
                        </div>
                      </Droppable>
                    </ConnectFunction>
                  </div>
                </div>
              </Card>
            </div>
          </Col>
          <Col
            xs={24}
          >
            <div
              className="ant-col ant-col-xs-24"
              style={Object {}}
            >
              <Card
                bordered={true}
                title="Stop sequence"
              >
                <div
                  className="ant-card ant-card-bordered"
                >
                  <div
                    className="ant-card-head"
                    style={Object {}}
                  >
                    <div
                      className="ant-card-head-wrapper"
                    >
                      <div
                        className="ant-card-head-title"
                      >
                        Stop sequence
                      </div>
                    </div>
                  </div>
                  <div
                    className="ant-card-body"
                    style={Object {}}
                  >
                    <div
                      style={
                        Object {
                          "height": "370px",
                          "overflowY": "auto",
                        }
                      }
                    >
                      <ConnectFunction
                        direction="vertical"
                        droppableId="trajekt"
                        getContainerForClone={[Function]}
                        ignoreContainerClipping={false}
                        isCombineEnabled={false}
                        isDropDisabled={false}
                        mode="standard"
                        renderClone={null}
                        type="DEFAULT"
                      >
                        <Droppable
                          direction="vertical"
                          droppableId="trajekt"
                          getContainerForClone={[Function]}
                          ignoreContainerClipping={false}
                          isCombineEnabled={false}
                          isDropDisabled={false}
                          mode="standard"
                          placeholder={null}
                          renderClone={null}
                          shouldAnimatePlaceholder={false}
                          snapshot={
                            Object {
                              "draggingFromThisWith": null,
                              "draggingOverWith": null,
                              "isDraggingOver": false,
                              "isUsingPlaceholder": false,
                            }
                          }
                          type="DEFAULT"
                          updateViewportMaxScroll={[Function]}
                          useClone={null}
                        >
                          <div
                            className="droppable-col"
                            data-rbd-droppable-context-id="0"
                            data-rbd-droppable-id="trajekt"
                          >
                            <AnimateInOut
                              on={null}
                              shouldAnimate={false}
                            />
                          </div>
                        </Droppable>
                      </ConnectFunction>
                    </div>
                  </div>
                </div>
              </Card>
            </div>
          </Col>
        </Provider>
      </App>
    </ErrorBoundary>
  </DragDropContext>
</Memo(DragDrop)>
`;
